Addition of MobileNet to application

titu1994 opened the pull request:
Adds the recently published MobileNets (all 16 variants, with weights trained on Imagenet).
Some comments : https://github.com/fchollet/keras/issues/6968

Inclusions : 
- MobileNet builder (alpha and imagesize allow the creation of all 16 types of mobile nets - (4 different sizes [128, 160, 192, 224] x 4 different alpha parameters [0.25, 0.50, 0.75, 1.00])
- DepthwiseConvolution2D layer
- depthwize_conv2d backend function (implemented in TF, raises exception in TH, left out CNTK since neither depthwise nor separable convolution backend functions were present)
- relu6 activation
- support for 1000 class predictions from imagenet
- tests for the application, imagenet_utils, layer, activation and backend function

Some issues - 
- ~~Weights are only available for data format "channels_last" at the moment. I have tried converting it to "channels_first", but it produces consistent garbage predictions. However, the model can be trained from scratch on either data format.~~
- ~~Model / Weights are for 1001 classes. This causes issues with decode_predictions which only supports 2 or 1000 classes. Therefore I drop the last class (which would usually be used for fine tuning a new class).~~

titu1994 commented:
Sorry, for so many commits and Travis runs.

fchollet commented:
The issues are easy to fix:

- Either the first or last entry in the 1001-large output is useless and can safely be dropped, resulting in a 1000-class net that will work with our existing class decoding utilities. It's probably the first entry.
- Any net that works for `channels_last` can be converted to `channels_first`. You just need to transpose the convolution kernels and shuffle the first Dense layer.

titu1994 commented:
@fchollet Thank you for the fixes. The 1st issue will be a weight update, so I will roll back the fix for decode_predictions in a future commit and update the weights on my end.

About the second issue, I attempted that. These models do not have a Dense layer, and only 1 final classification layer using pointwise convolutions and softmax. 

Also, transposing the weights is not possible, since DepthwiseConv2D is just the 1st part of SeparableConv2D. It does not support Theano, and even Xception models have weights only for tf_dim_ordering, which is equivalent to channels_last, and changes the data format internally to handle this.

Should I take the same approach?

titu1994 commented:
@fchollet Just an update. I have updated the model to use 1000 classes, and updated the weights on my end to support 1000 classes (by dropping the 0th class from the 1001 class weights). 

The performance of mobilenet_1_0_224 is just 1 percentage points below Xception on the test image in my repository at the moment.

I wait for your comment, about whether to handle channels_first case in the same manner as Xception, before updating the code.

titu1994 commented:
The build errors are due to TimeDistributed errors in CNTK and Theano, and one pip install error for Integration Tests.

titu1994 commented:
Tests are failing due to time distributed wrapper, not because of this PR.

titu1994 commented:
Restarting tests after https://github.com/fchollet/keras/pull/7018

titu1994 commented:
@fchollet All tests passing now. I have updated the MobileNet models to enforce channels_last data format everywhere, to match Xception.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+'''MobileNet models for Keras.
```
Add line break

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+# Reference
+- [MobileNets: Efficient Convolutional Neural Networks for
+   Mobile Vision Applications](https://arxiv.org/pdf/1704.04861.pdf))
+'''
```
Add information about which MobileNet models can be built and what their ImageNet test performance is (and parameter count)

fchollet replied:
Also add info about where the weights came from, for licensing purposes

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+from keras.layers.core import Activation, Dropout, Reshape
+from keras.activations import relu
+from keras.layers.convolutional import Convolution2D, DepthwiseConvolution2D
+from keras.layers.pooling import GlobalAveragePooling2D, GlobalMaxPooling2D
```
Don't import submodules like `pooling` they are no longer meant to be used. Also avoid importing specific layers. Instead import `layers`

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+
+import warnings
+
+from keras.models import Model
```
This is a Keras submodule, don't import from Keras, use a relative import

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+def MobileNet(input_shape=None, alpha=1.0, depth_multiplier=1,
+              dropout=1e-3, include_top=True, weights='imagenet',
+              input_tensor=None, pooling=None, classes=1000):
+    ''' Instantiate the MobileNet architecture.
```
Remove leading space.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+        therefore it only works with the data format
+        `image_data_format='channels_last'` in your Keras config
+        at ~/.keras/keras.json.
+        # Arguments
```
Add line break above

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+        Note that only TensorFlow is supported for now,
+        therefore it only works with the data format
+        `image_data_format='channels_last'` in your Keras config
+        at ~/.keras/keras.json.
```
Put code delimiter around path

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+            classes: optional number of classes to classify images
+                into, only to be specified if `include_top` is True, and
+                if no `weights` argument is specified.
+        # Returns
```
Add line break above

titu1994 replied:
Fixed.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+        '''
+
+    if K.backend() != 'tensorflow':
+        raise AttributeError('Only Tensorflow backend is currently supported, '
```
Should this be `RuntimeError`?

titu1994 replied:
Yes it should. I consider `tensorflow` an attribute value for `backend`, therefore the Attribute error. 
Fixed.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+                if no `weights` argument is specified.
+        # Returns
+            A Keras model instance.
+        '''
```
Needs a `Raises` section

titu1994 replied:
Fixed.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+
+def MobileNet(input_shape=None, alpha=1.0, depth_multiplier=1,
+              dropout=1e-3, include_top=True, weights='imagenet',
+              input_tensor=None, pooling=None, classes=1000):
```
One keyword argument per line

Why is dropout an argument here? And why so low?

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+                It should have exactly 3 inputs channels,
+                and width and height should be no smaller than 32.
+                E.g. `(200, 200, 3)` would be one valid value.
+            alpha: width multiplier of the MobileNet.
```
Explain what it is, most people are not familiar with "width multiplier"

titu1994 replied:
Added a small explanation.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+    if weights == 'imagenet':
+        assert depth_multiplier == 1, "If imagenet weights are being loaded, depth multiplier must be 1"
+
+        assert alpha in [0.25, 0.50, 0.75, 1.0], "If imagenet weights are being loaded, alpha can be one of" \
```
Line too long
Don't use `\` for breaking lines
Use a ValueError instead of an assert

titu1994 replied:
Fixed.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+        rows = int(input_shape[rows])
+        cols = int(input_shape[cols])
+
+        assert rows == cols and rows in [128, 160, 192, 224], "If imagenet weights are being loaded," \
```
Same

titu1994 replied:
Fixed.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+        else:
+            img_input = input_tensor
+
+    x = __create_mobilenet(classes, img_input, include_top, alpha, depth_multiplier, dropout, pooling)
```
Use a single leading underscore to denote a private function

titu1994 replied:
Fixed.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,275 @@
+            raise AttributeError('Weights for Channels Last format are not available')
+
+        if alpha == 1.0:
+            alpha_text = "1_0"
```
Make consistent use of `'` and '`"` (always use `'`)

titu1994 replied:
Fixed.

fchollet commented on keras/backend/theano_backend.py:
```diff
@@ -1904,6 +1904,11 @@ def separable_conv2d(x, depthwise_kernel, pointwise_kernel, strides=(1, 1),
     raise NotImplementedError
 
 
+def depthwise_conv2d(x, depthwise_kernel, strides=(1, 1), padding='valid',
```
Also add to CNTK backend

titu1994 replied:
Added. I think the same should be done for separable_conv2d as well.

fchollet replied:
Yes, that's right

fchollet commented on keras/layers/convolutional.py:
```diff
@@ -1032,6 +1032,198 @@ def get_config(self):
+        or 4D tensor with shape:
+        `(batch, new_rows, new_cols, filters)` if data_format='channels_last'.
+        `rows` and `cols` values might have changed due to padding.
+    """
```
Needs a `Raises` section
Line break after each section

titu1994 replied:
The layer doesn't raise any error other than the standard ValueErrors (for input shape of rank < 4 and channel dimention not being defined). 

I see that other layers (especially SeparableConvolution2D) also do not raise any error other than the two above. Should I raise a RuntimeError when using backends other than tensorflow?

fchollet commented on keras/layers/convolutional.py:
```diff
@@ -1032,6 +1032,198 @@ def get_config(self):
+                                     padding=self.padding,
+                                     dilation_rate=self.dilation_rate,
+                                     data_format=self.data_format)
+
```
Unnecessary space

fchollet commented on keras/applications/imagenet_utils.py:
```diff
@@ -58,7 +58,7 @@ def decode_predictions(preds, top=5):
         raise ValueError('`decode_predictions` expects '
                          'a batch of predictions '
-                         '(i.e. a 2D array of shape (samples, 1000)). '
+                         '(i.e. a 2D array of shape (samples, 1000).'
```
Def need the extra ')` (closing)

titu1994 commented:
Added most of the corrections. 

A few such as adding a "Raises" section to the DepthwiseConv2D layer and adding dropout as a parameter needs to be discussed, but other than those two, all corrections are applied.

titu1994 commented:
@fchollet I've made the corrections as requested.

fchollet commented:
Can someone please help review?

titu1994 commented:
@fchollet Had to add `relu6` to `activations.py`.

Initially, I was using python lambda to use a relu with max value of 6. However, upon testing model.save(...) and then load_model(..), I found that models which are fine tuned using such Activations do not de-serialize properly.

The easiest fix is adding `relu6` to activations, and make it directly use the `max_value` parameter of the original K.relu(...) function. This allows for proper serialization and de-serialization.

titu1994 commented:
@fchollet Since there doesn't seem to be other reviewers, perhaps you can mention some general guidelines to follow, to which I can adhere to? 

I tried to match Keras documentation styles, and code style from other layers, tests, and applications as much as possible. The only place left which doesn't strictly adhere to this are the private functions that build the model (generally the code is inlined in the method, here I have used a few different methods).

ahundt commented:
I think relu6 should just be defined locally in the application file, or it should be generalized unless 6 is a fundamentally better max for many wide use cases. Plus, I think it is possible to pass relu6 as a param to the Activation, it doesn't need to be a string. 

I've been keeping an eye on this and overall the code structure/syntax looks fine to me but unfortunately I can't say I've carefully compared this to the paper...

titu1994 commented:
@ahundt If you mean relu6 should be local to the file in the following manner : 
```
import keras.activations

def relu6(x):
    return K.relu(x, max_value=6)

keras.activations.relu6 = relu6
... 
x = Activation('relu6', ...)(x)
```
Yes, the above works.

However, using a lambda to wrap relu with max_value parameter does not work. It will not deserialize correctly with load_model. Unfortunately, I have found no way of passing arbitrary arguments to Activation layer (so that it can defer them to the actual activation function). It also makes very little sense, given that `relu` is the only 

I added relu6 to Keras, because Tensorflow has a specific activation function called `relu6` [(tf.nn.relu6)](https://www.tensorflow.org/api_docs/python/tf/nn/relu6). Since its inclusion will not affect the Theano or CNTK backends, it made more sense to me to incorporate it into the activations file and be in line with the Tensorflow API. If @fchollet decides that the above mentioned fix is sufficient for this purpose, I will roll back this change.

As to comparison with the paper, the model is correct in its implementation. The tensorflow graphs match exactly. Further proof is that the tensorflow provided weights, upon translation to Keras, provide exactly same predictions as Tensorflow with same accuracy.

ahundt commented:
Sound reasoning, thanks for the explanation! You always do good work.

fchollet commented:
> the above works

I'd recommend doing that.

> I added relu6 to Keras, because Tensorflow has a specific activation function called relu6

A legacy of patterns used in early DL models at Google. Including it in TF was a mistake IMO. It's technical debt. No need to burden the Keras API with it.

titu1994 commented:
@fchollet Got it. Rolled back the `relu6` activation and applied the equivalent fix.

fchollet commented:
If the layer (and relu6) turn out to be widely used in the future, we can then easily move them from `mobilenet.py` to the code codebase.

titu1994 commented on keras/layers/convolutional.py:
```diff
@@ -1032,6 +1032,205 @@ def get_config(self):
+        `(batch, new_rows, new_cols, filters)` if data_format='channels_last'.
+        `rows` and `cols` values might have changed due to padding.
+
+    # Raises
```
What should it say? This layer does not raise any specific error other than ValueError for input shape not being a 4D tensor or channel dimention of input shape not being defined.

Both of these are from SeparableConv2D, and that layer does not have a Raises section.

fchollet replied:
If there are no errors to report, do not include the section header.

titu1994 commented:
@fchollet I agree with your assessment. 

I had to make a few alterations to `__init__.py` of the applications folder to expose the DepthwiseConv2D / DepthwiseConvolution2D layers and relu6 activation. Also, the convolution tests must now import DepthwiseConv2D from keras.applications.mobilenet rather than keras.layers, but other than that no major modifications.

The other corrections have been applied as suggested.

titu1994 commented:
@fchollet Seems tests fail because their is a deserialization test in the `layer_test` which cannot deserialize the layer from config. 

`recovered_model = Model.from_config(model_config)` (Line 100 in test_utils.py) does take custom objects, but the `layer_test` does not have an argument to pass the custom object to that section of the test.

I can modify the layer_test to take an optional argument `custom_objects=None` at the end, and then pass that to the respective lines (the cleanest solution I can think of, and I would prefer this if allowed).

If not that, I would have to drop the tests for the layer (a last resort option).

fchollet commented:
You can use a custom object scope to avoid having to pass around custom objects: https://keras.io/utils/#customobjectscope

fchollet commented on keras/applications/__init__.py:
```diff
@@ -3,3 +3,7 @@
+from .mobilenet import MobileNet
+from .mobilenet import DepthwiseConv2D
+from .mobilenet import DepthwiseConvolution2D
+from .mobilenet import relu6
```
These 3 should not be exposed as part of the `applications` module (only as part of the `mobilenet` submodule)

fchollet commented on tests/keras/layers/convolutional_test.py:
```diff
@@ -276,6 +277,47 @@ def test_separable_conv_2d():
 
+@pytest.mark.skipif(K.backend() != 'tensorflow', reason='Requires TF backend')
+@keras_test
+def test_depthwise_conv_2d():
```
This test should be in `applications_test`

titu1994 commented:
@fchollet Thank you for the suggestions. Learned something new with CustomObjectScope. 

Applied the fix and tests pass locally now, and hopefully will pass on Travis as well.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,572 @@
+
+class DepthwiseConv2D(Conv2D):
+    """
+    Depthwise separable 2D convolution.
```
Put this line on the first line of the docstring

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,572 @@
+    return x
+
+
+def _create_mobilenet(classes, img_input, include_top, alpha, depth_multiplier, dropout, pooling):
```
Please don't create an additional function that is only used once. Put this code in the model constructor (line 419)

ahundt replied:
I found the design here much more practical than the design of other classes in the applications folder implementations. My situations are slightly different so I'm able to just call this internal function. In the applications of other utilities such as [ResNet50](https://github.com/fchollet/keras/blob/master/keras/applications/resnet50.py#L125) I've had to manually copy/paste the code or otherwise rewrite things to actually use it. 

In other words, for models there are very specific and valid reasons why such `_create*()` functions should exist.

fchollet replied:
Either it's part of the public API of the module, either its existence is just an implementation detail. Having to copy internal model code is not a bad pattern. In general, if a function not part of the public API is used exactly once, then it should not exist.

ahundt replied:
> Either it's part of the public API of the module, either its existence is just an implementation detail.

I'd be very happy if the portion on which typical fine tuning techniques are applied were easy to create and load as part of the public API, i.e. this function.

> Having to copy internal model code is not a bad pattern.

Many of the principles you follow are truly excellent, but according to software engineering best practices and current research, copying code is a pattern with .

- [Best Practices for Scientific Computing. Wilson G et al](http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1001745&utm_content=buffer184f2&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer). 
    - [Don't repeat yourself](http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1001745&utm_content=buffer184f2&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer#s5)
- [Do Code Clones Matter? Juergens et al](https://arxiv.org/abs/1701.05472)

titu1994 replied:
@ahundt I generally agree to what you are saying. There are some very finicky models (like DenseNets and DenseNetFCNs) which really require their own separate methods, simply because of the large variety of very similar models that can be built due to complex hyper-parameters which interact and conflict with one another. 

This however is a special case, where the model is completely linear and doesn't even have residual connections. It doesn't really hurt much to inline it.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,572 @@
+
+# Aliases
+
+DepthwiseConvolution2D = DepthwiseConv2D
```
No need, we don't mean to expert this layer.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,572 @@
+    else:
+        inputs = img_input
+    # Create model.
+    model = Model(inputs, x, name='mobilenet')
```
Maybe format the name with alpha and rows? Up to you.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,572 @@
+
+
+def _conv_block(input, filters, alpha, kernel=(3, 3), strides=(1, 1)):
+    """ Adds an initail convolution layer (with batch normalization and relu6)
```
"Initial". Please add a properly formatted docstring.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,572 @@
+def _depthwise_conv_block(input, pointwise_conv_filters, alpha,
+                          depth_multiplier=1, strides=(1, 1), id=1):
+    """ Adds a depthwise convolution block (depthwise conv, batch normalization, relu6,
+        pointwise convolution, batch normalization and relu6)
```
Same, please add a properly formatted docstring.

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,572 @@
+            and width and height should be no smaller than 32.
+            E.g. `(200, 200, 3)` would be one valid value.
+        alpha: controls the width of the network.
+            If `alpha` < 1.0, proportionally decreases the number
```
Use list markers to format this list

fchollet commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,572 @@
+                         ' as true, `classes` should be 1001')
+
+    if weights == 'imagenet':
+        assert depth_multiplier == 1, 'If imagenet weights are being loaded, ' \
```
Don't use `assert`, raise a ValueError. Don't break lines with `\`

titu1994 commented:
@fchollet Made the corrections as requested. Sorry about the documentation for the private functions, I assumed they would not need full docstrings since they are private.

I have a habit of having builder methods, especially because there are some very complex model builder functions for DenseNets and DenseNetFCNs, which would not look great if it was inlined into an already long method which the user calls. In either case, this model is simple enough to inline.

ahundt commented on keras/applications/mobilenet.py:
```diff
@@ -495,78 +607,3 @@ def _depthwise_conv_block(input, pointwise_conv_filters, alpha,
     return x
-
-
-def _create_mobilenet(classes, img_input, include_top, alpha, depth_multiplier, dropout, pooling):
```
I don't want to delay the merge of this code for this discussion. My comment was hidden in the previous review stage.

As I mentioned in https://github.com/fchollet/keras/pull/7009#discussion_r124160557, the earlier design here would be more effective as part of the public API. I base this assertion on the following research:



- [Best Practices for Scientific Computing. Wilson G et al](http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1001745&utm_content=buffer184f2&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer). 
    - [Don't repeat yourself](http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1001745&utm_content=buffer184f2&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer#s5)
- [Do Code Clones Matter? Juergens et al](https://arxiv.org/abs/1701.05472)

We should make sure code cloning is not necessary when it is easy, clear, and concise to do so.

I can also provide reference to many repositories where copying of the code is occurring.

titu1994 replied:
Ah, I responded to that. To re-iterate : 
I generally agree to what you are saying. There are some very finicky models (like DenseNets and DenseNetFCNs) which really require their own separate methods, simply because of the large variety of very similar models that can be built due to complex hyper-parameters which interact and conflict with one another.

This however is a special case, where the model is completely linear and doesn't even have residual connections. It doesn't really hurt much to inline it.

titu1994 commented:
@fchollet Thank you for all of your comments and discussion!

fchollet commented:
@titu1994 thank you for the contribution 👍

fchollet commented:
`mobilenet.py` had several bugs, which I have fixed in this commit: https://github.com/fchollet/keras/commit/8040ad72dd488feac9843caca4f875186d3b75c1
(in particular, the default configuration would crash, `int` was taken for `None` values, `input_shape` was referred to before being inferred...)

Additionally, at this time, it appears that none of the weight files produce accurate results -- all results appear to be random. Please take a look. If it can be fixed within a reasonable timeframe, I'll remove this application.

titu1994 commented:
@fchollet Sorry about the input_shape bug. I'm surprised it passed the tests without crashing.

I confirmed correct predictions with elephant and giraffe, but after testing it on a cheetah it is giving wrong predictions as wolf so I can confirm there is something wrong with the weights. Will look into the weights right away.

fchollet commented:
The tests didn't test the default, they tested `weights=None`, with or without pooling, with or without include_top. It's the weight loading configs that were failing.

I only tested with 2 different images and got random results for every imagenet config (different alphas). Interestingly, I get completely different results for each config.

titu1994 commented:
Just ran the model and the checkpoint on the same image. Both are giving the exact same results. 
Here is a gist to execute the keras (master version with above mentioned fix) and the tensorflow checkpoint: https://gist.github.com/titu1994/c110987f5ac6a0f1e3b64320764b1f40

Results (Cheetah - TF checkpoint): 
```
D:\Users\Yue\Anaconda3\python.exe D:/Users/Yue/PycharmProjects/MobileNets/_weight_extraction/predict_checkpoint.py
Using TensorFlow backend.
2017-06-30 14:47:24.960820: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:47:24.961092: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE2 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:47:24.961325: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:47:24.961558: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:47:24.961792: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:47:24.962073: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:47:24.962308: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:47:24.962541: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:47:25.598830: I c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\common_runtime\gpu\gpu_device.cc:940] Found device 0 with properties: 
name: GeForce GTX 980M
major: 5 minor: 2 memoryClockRate (GHz) 1.1265
pciBusID 0000:01:00.0
Total memory: 4.00GiB
Free memory: 3.32GiB
2017-06-30 14:47:25.599106: I c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\common_runtime\gpu\gpu_device.cc:961] DMA: 0 
2017-06-30 14:47:25.599241: I c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\common_runtime\gpu\gpu_device.cc:971] 0:   Y 
2017-06-30 14:47:25.599405: I c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\common_runtime\gpu\gpu_device.cc:1030] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 980M, pci bus id: 0000:01:00.0)

Predicted: [[('n02130308', 'cheetah', 0.28019959), ('n02117135', 'hyena', 0.25728428), ('n02423022', 'gazelle', 0.097853415), ('n02437616', 'llama', 0.094322421), ('n02099429', 'curly-coated_retriever', 0.090131752)]]

Process finished with exit code 0
```

Results (Cheetah - Keras master)
```
D:\Users\Yue\Anaconda3\python.exe D:/Users/Yue/PycharmProjects/MobileNets/predict_imagenet.py
Using TensorFlow backend.
2017-06-30 14:53:35.194721: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:53:35.194980: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE2 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:53:35.195221: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:53:35.195905: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:53:35.196206: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:53:35.196461: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:53:35.196634: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:53:35.196867: W c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
2017-06-30 14:53:35.815831: I c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\common_runtime\gpu\gpu_device.cc:940] Found device 0 with properties: 
name: GeForce GTX 980M
major: 5 minor: 2 memoryClockRate (GHz) 1.1265
pciBusID 0000:01:00.0
Total memory: 4.00GiB
Free memory: 3.32GiB
2017-06-30 14:53:35.816151: I c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\common_runtime\gpu\gpu_device.cc:961] DMA: 0 
2017-06-30 14:53:35.816278: I c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\common_runtime\gpu\gpu_device.cc:971] 0:   Y 
2017-06-30 14:53:35.816418: I c:\tf_jenkins\home\workspace\release-win\m\windows-gpu\py\35\tensorflow\core\common_runtime\gpu\gpu_device.cc:1030] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 980M, pci bus id: 0000:01:00.0)

Predicted: [[('n02130308', 'cheetah', 0.99743026), ('n02128385', 'leopard', 0.0010753422), ('n02129165', 'lion', 0.00069186132), ('n02128757', 'snow_leopard', 0.00059767498), ('n02127052', 'lynx', 0.00012871811)]]
```

Results (African Elephant - TF checkpoint)
```
D:\Users\Yue\Anaconda3\python.exe D:/Users/Yue/PycharmProjects/MobileNets/_weight_extraction/predict_checkpoint.py
# TF debug statements again

Predicted: [[('n02130308', 'cheetah', 0.28019959), ('n02117135', 'hyena', 0.25728428), ('n02423022', 'gazelle', 0.097853415), ('n02437616', 'llama', 0.094322421), ('n02099429', 'curly-coated_retriever', 0.090131752)]]

```

Results (African Elephant - Keras master)
```
D:\Users\Yue\Anaconda3\python.exe D:/Users/Yue/PycharmProjects/MobileNets/predict_imagenet.py
# TF debug statements again

Predicted: [[('n02130308', 'cheetah', 0.28019962), ('n02117135', 'hyena', 0.25728455), ('n02423022', 'gazelle', 0.097853422), ('n02437616', 'llama', 0.094322428), ('n02099429', 'curly-coated_retriever', 0.090131581)]]

```

titu1994 commented:
I got the error. It was the wrong weight normalization.

titu1994 commented:
@fchollet  To fix, the import for preprocess_input should be changed from 
`from keras.applications.imagenet_utils import preprocess_input`

to

`from keras.applications.inception_v3 import preprocess_input`

fchollet commented:
I can confirm this fixes it, thanks for looking into it.

Additionally, we need to figure out a way to test applications under all configurations (including weights) without having to redownload everything 20 times a day in CI.

titu1994 commented:
@fchollet I agree. More than that, I think there must be someway for CI runs to be canceled before execution. This would help in cases where there are 2 or more commits in a short time frame (within 30 minutes of each other) from the same PR.

I am sadly guilty of having done this during this PR, and many of those CI runs could have been avoided.

fchollet commented:
That is already the case. We configured it a couple weeks ago.

titu1994 commented:
@fchollet Oh, that is fantastic.

clarle commented on keras/applications/mobilenet.py:
```diff
@@ -0,0 +1,616 @@
+        x = GlobalAveragePooling2D()(x)
+        x = Reshape(shape, name='reshape_1')(x)
+        x = Dropout(dropout, name='dropout')(x)
+        x = Convolution2D(classes, (1, 1), padding='same', name='conv_preds')(x)
```
@titu1994 I was reading the MobileNets paper, and I was curious why you decided to use a Conv2D layer here instead of a Dense layer, since this layer is supposed to be the fully-connected one.

I know in the end the weights will make the Conv2D layer equivalent to a fully-connected one, but is there any particular reason you used Conv2D here instead of Dense?  Just wanted to better improve my understanding, thanks so much!

zeruniverse commented:
Is there any reason that `channels_first` for mobilenet is not supported other than issues related to ImageNet weights?

I removed everything about imagenet weights and `channels_last` restrictions and trained a `channels_first` network from scratch but the network just gives random guesses (no errors). If I train a `channels_last` network from scratch, the results look reasonable.
